File: ios/GhostChat/Core/Files/FileCatalog.swift
import Foundation

/// Static registry of MIME types Ghost Chat accepts as attachments. Used to
/// reject unsupported payloads at the UI layer and to guess a MIME type from a
/// filename when the OS picker didn't provide one.
///
/// Mirror of Android `core/files/FileCatalog.kt`. Keep the table in sync.
enum FileCatalog {

    enum Category: String, Equatable {
        case image
        case video
        case audio
        case document
    }

    struct Entry: Equatable {
        let mimeType: String
        let ext: String
        let category: Category
    }

    static let entries: [Entry] = [
        // Images
        .init(mimeType: "image/jpeg", ext: "jpg",  category: .image),
        .init(mimeType: "image/jpeg", ext: "jpeg", category: .image),
        .init(mimeType: "image/png",  ext: "png",  category: .image),
        .init(mimeType: "image/gif",  ext: "gif",  category: .image),
        .init(mimeType: "image/heic", ext: "heic", category: .image),
        .init(mimeType: "image/webp", ext: "webp", category: .image),
        // Video
        .init(mimeType: "video/mp4",       ext: "mp4", category: .video),
        .init(mimeType: "video/quicktime", ext: "mov", category: .video),
        // Audio
        .init(mimeType: "audio/mpeg", ext: "mp3", category: .audio),
        .init(mimeType: "audio/mp4",  ext: "m4a", category: .audio),
        .init(mimeType: "audio/aac",  ext: "aac", category: .audio),
        .init(mimeType: "audio/wav",  ext: "wav", category: .audio),
        // Documents
        .init(mimeType: "application/pdf",   ext: "pdf",  category: .document),
        .init(mimeType: "application/msword", ext: "doc", category: .document),
        .init(
            mimeType: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            ext: "docx", category: .document
        ),
        .init(mimeType: "text/plain",      ext: "txt", category: .document),
        .init(mimeType: "application/zip", ext: "zip", category: .document)
    ]

    static func isSupportedMimeType(_ mime: String) -> Bool {
        entries.contains { $0.mimeType == mime }
    }

    static func categoryFor(mimeType: String) -> Category? {
        entries.first { $0.mimeType == mimeType }?.category
    }

    static func mimeType(forFilename filename: String) -> String? {
        let ext = (filename as NSString).pathExtension.lowercased()
        return entries.first { $0.ext == ext }?.mimeType
    }

    /// Primary (first-listed) extension for a MIME type. `.jpg` wins over `.jpeg`.
    static func primaryExtension(forMimeType mime: String) -> String? {
        entries.first { $0.mimeType == mime }?.ext
    }
}

File: android/app/src/main/java/com/kordar/ghostchat/core/files/FileCatalog.kt
package com.kordar.ghostchat.core.files

/**
 * Static registry of MIME types Ghost Chat accepts as attachments. Used to
 * reject unsupported payloads at the UI layer and to guess a MIME type from a
 * filename when the OS picker didn't provide one.
 *
 * Mirror of iOS `Core/Files/FileCatalog.swift`. Keep the table in sync.
 */
object FileCatalog {

    enum class Category { IMAGE, VIDEO, AUDIO, DOCUMENT }

    data class Entry(val mimeType: String, val ext: String, val category: Category)

    val entries: List<Entry> = listOf(
        // Images
        Entry("image/jpeg", "jpg",  Category.IMAGE),
        Entry("image/jpeg", "jpeg", Category.IMAGE),
        Entry("image/png",  "png",  Category.IMAGE),
        Entry("image/gif",  "gif",  Category.IMAGE),
        Entry("image/heic", "heic", Category.IMAGE),
        Entry("image/webp", "webp", Category.IMAGE),
        // Video
        Entry("video/mp4",       "mp4", Category.VIDEO),
        Entry("video/quicktime", "mov", Category.VIDEO),
        // Audio
        Entry("audio/mpeg", "mp3", Category.AUDIO),
        Entry("audio/mp4",  "m4a", Category.AUDIO),
        Entry("audio/aac",  "aac", Category.AUDIO),
        Entry("audio/wav",  "wav", Category.AUDIO),
        // Documents
        Entry("application/pdf",   "pdf",  Category.DOCUMENT),
        Entry("application/msword", "doc", Category.DOCUMENT),
        Entry(
            "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            "docx", Category.DOCUMENT
        ),
        Entry("text/plain",      "txt", Category.DOCUMENT),
        Entry("application/zip", "zip", Category.DOCUMENT)
    )

    fun isSupportedMimeType(mime: String): Boolean =
        entries.any { it.mimeType == mime }

    fun categoryFor(mimeType: String): Category? =
        entries.firstOrNull { it.mimeType == mimeType }?.category

    fun mimeType(filename: String): String? {
        val ext = filename.substringAfterLast('.', "").lowercase()
        if (ext.isEmpty()) return null
        return entries.firstOrNull { it.ext == ext }?.mimeType
    }

    /** Primary (first-listed) extension for a MIME type. `.jpg` wins over `.jpeg`. */
    fun primaryExtension(mime: String): String? =
        entries.firstOrNull { it.mimeType == mime }?.ext
}
